File: web/admin-console/tests/oversight.test.mjs
// oversight.test.mjs is the console half of P26 wave 26e.
//
// The platform half (internal/adminops/oversight_test.go) asserts the read model's absences. This file
// asserts what the screen does with them: that a missing legal record does not render as an empty
// acceptance table, that an unread readiness surface does not render as "not configured", that an
// unknown version is not shown as a version, and that the payments gap renders as a statement rather
// than as a page with a zero on it.

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import { join } from "node:path";

const ROOT = new URL("..", import.meta.url).pathname;
const read = (rel) => readFile(join(ROOT, rel), "utf8");

test("🔴 6.1 — a session shows its verified factor, and the fixture IdP is not dressed as production", async () => {
  const page = await read("src/app/oversight/page.tsx");
  assert.match(page, /s\.factor/, "the session's factor is not rendered");
  assert.match(page, /s\.multi_factor/, "single-factor and multi-factor sessions are not distinguishable");
  assert.match(page, /s\.verified_at/, "the verification time is not rendered");
  assert.match(page, /identity_provider\.test_mode/, "the page does not render whether the IdP is the fixture");
  assert.match(page, /identity_provider\.note/, "the page does not render the identity provider's own caveat");
  // And no token or factor VALUE has a field to travel in.
  const types = await read("src/lib/types.ts");
  const sessionRow = types.match(/export type SessionRow = {([^}]*)}/);
  assert.ok(sessionRow, "SessionRow is not declared");
  for (const field of ["token", "secret", "code", "assertion"]) {
    assert.equal(
      new RegExp(`\\b${field}\\b`, "i").test(sessionRow[1]),
      false,
      `SessionRow carries a \`${field}\` field — the factor NAME is what a reviewer needs, never its value`,
    );
  }
});

test("🔴 6.2 — an accepted version links the ARCHIVED text at its content hash", async () => {
  const page = await read("src/app/oversight/page.tsx");
  assert.match(page, /l\.archive_href/, "the archived text is not linked");
  assert.match(page, /l\.accepted_hash/, "the accepted content hash is not rendered");
  assert.match(page, /l\.owed_version/, "versions owed after a material publication are not rendered");
  assert.match(page, /nothing owed/, "a tenant that owes nothing is not stated as such");
  // A missing legal record must not render as an empty acceptance table.
  assert.match(page, /!view\.legal_known/, "an absent legal record is not branched on");
  assert.match(
    page.replace(/\s+/g, " "),
    /would read as 'nobody owes anything'/,
    "the page does not say why an empty acceptance table would be a different claim",
  );
});

test("🔴 6.3 — an integration has three states, and an unread readiness surface is not 'not configured'", async () => {
  const types = await read("src/lib/types.ts");
  assert.match(
    types,
    /export type IntegrationState = "absent" \| "configured" \| "degraded";/,
    "IntegrationState is not three values — a boolean would call a decision a fault or a fault a decision",
  );
  const page = await read("src/app/oversight/page.tsx");
  const tones = page.match(/const INTEGRATION_TONE[^=]*=\s*{([^}]*)}/);
  assert.ok(tones, "the page declares no integration tone map");
  // `absent` is NEUTRAL: nothing is wrong, nothing is configured. Only `degraded` earns the hazard.
  assert.match(tones[1], /absent:\s*"neutral"/, "an unconfigured integration is painted as a hazard");
  assert.match(tones[1], /degraded:\s*"warn"/, "a degraded integration is not marked as a fault");
  assert.match(page, /!view\.integrations_known/, "an unread readiness surface is not branched on");
  assert.match(
    page.replace(/\s+/g, " "),
    /'nothing is configured' and 'we did not ask' are different answers/,
    "the page does not say why an unread readiness surface is not the same as 'not configured'",
  );
  assert.match(page, /n\.failure_class/, "a degraded integration does not name its failure class");
});

test("🔴 6.4 — an unknown deployed version is stated, never shown as a version", async () => {
  const page = await read("src/app/oversight/page.tsx");
  assert.match(page, /d\.unknown \?/, "the unknown case is not branched on");
  assert.match(page, /requires \$\{d\.missing_collection\}/, "an unknown version does not name what would make it readable");
  // No inference vocabulary anywhere.
  for (const word of ["estimated", "approximately", "inferred from", "probably"]) {
    assert.equal(
      new RegExp(word, "i").test(page.replace(/is inferred from an API contract version[^*]*/g, "")),
      false,
      `the oversight page offers an ${word} version — an inferred version rendered as a version is a ` +
        `wrong number that gets acted on during an incident`,
    );
  }
});

test("🔴 6.5 — the payments gap renders as a statement, with no count and no zero", async () => {
  const page = await read("src/app/oversight/page.tsx");
  assert.match(page, /not_yet_readable\.map/, "the not-yet-readable block is not rendered");
  assert.match(page, /n\.requires/, "the missing collection is not rendered");
  assert.match(page, /n\.statement/, "the statement is not rendered");
  // No table, no count, no Num in that panel: the whole point is that there is nothing to count.
  const panel = page.slice(page.indexOf('id: "not-yet"'));
  assert.equal(/<DataTable/.test(panel), false, "the not-yet-readable panel renders a table");
  assert.equal(/<Num /.test(panel), false, "the not-yet-readable panel renders a figure");
});

test("🔴 6.6 — the /oversight destination is registered, gated and in the nav", async () => {
  const surfaces = await read("src/lib/surfaces.ts");
  assert.match(surfaces, /capability: "audit\.read",\s*\n\s*href: "\/oversight"/,
    "the /oversight destination is not registered against audit.read");
  const page = await read("src/app/oversight/page.tsx");
  assert.match(page, /hasCapability\(identity, "audit\.read"\)/, "the page does not check the capability");
  assert.match(page, /<DeniedState/, "the page does not render a denial with an escalation path");
});

File: internal/adminops/oversight_test.go
package adminops_test

import (
	"encoding/json"
	"reflect"
	"strings"
	"testing"

	"github.com/heros-foreal/agentd/internal/adminops"
	"github.com/heros-foreal/agentd/internal/adminrbac"
)

// oversight_test.go covers P26 wave 26e.
//
// The property that matters most here is the one about ABSENCE. Three of these reads can be absent —
// the legal record, the readiness surface, the deployed version — and each absence has to be reported
// as itself rather than as an empty table, a `0`, or an `absent` state. "Nothing is configured" and
// "we did not ask" are opposite claims that render identically if nobody insists otherwise.

type testReadiness struct{}

func (testReadiness) Describe() string { return "test readiness surface" }

func (testReadiness) Integrations() []adminops.IntegrationRow {
	return []adminops.IntegrationRow{
		{Name: "metrics", State: adminops.IntegrationConfigured, Source: "platform readiness surface"},
		{Name: "error-monitoring", State: adminops.IntegrationAbsent, Source: "platform readiness surface"},
		{Name: "trace-export", State: adminops.IntegrationDegraded,
			FailureClass: "configured, endpoint unreachable for 14 consecutive flushes",
			Source:       "platform readiness surface"},
	}
}

func oversightView(t *testing.T, h *harness, cfg adminops.OversightConfig) adminops.OversightView {
	t.Helper()
	svc, err := adminops.NewOversightService(h.exec, cfg)
	if err != nil {
		t.Fatalf("NewOversightService: %v", err)
	}
	view, err := svc.View(h.ctx(adminrbac.RolePlatformSRE))
	if err != nil {
		t.Fatalf("View: %v", err)
	}
	return view
}

// TestAnOperatorSessionShowsTheFactorThatAuthenticatedIt defends P26 task 6.1.
//
// It runs against the REAL verifier — the harness's sessions were issued by
// `adminidentity.Authenticator` after a real assertion and a real MFA check — and the surface renders
// what that verifier recorded. It also asserts the surface does NOT represent the test-mode fixture as
// a production identity provider.
func TestAnOperatorSessionShowsTheFactorThatAuthenticatedIt(t *testing.T) {
	h := newHarness(t)
	view := oversightView(t, h, adminops.OversightConfig{
		Sessions: h.sessions, Identity: h.identityInfo(),
	})

	if len(view.Sessions) == 0 {
		t.Fatal("no operator sessions are shown, though the fixture signed four principals in")
	}
	for _, s := range view.Sessions {
		if s.Factor == "" {
			t.Fatalf("session %s shows no factor — a reviewer would be left inferring authentication "+
				"strength, which is what this surface exists to stop", s.SessionID)
		}
		if s.VerifiedAt == "" {
			t.Fatalf("session %s shows no verification time", s.SessionID)
		}
		if !s.MultiFactor {
			t.Fatalf("session %s is not distinguishable as multi-factor, though the fixture verified one",
				s.SessionID)
		}
	}
	// 🔴 The surface claims no real IdP. The verifier is real; the ISSUER is the fixture, and saying so
	// is the difference between a demo that is honest and one that is misleading.
	if !view.IdentityProvider.TestMode {
		t.Fatal("the fixture IdP is not reported as test mode")
	}
	if !strings.Contains(view.IdentityProvider.Note, "TEST-MODE") {
		t.Fatalf("the surface does not say the identity provider is the fixture: %q", view.IdentityProvider.Note)
	}
	if !strings.Contains(view.IdentityProvider.Note, "real one") {
		t.Fatalf("the surface does not say the VERIFIER is real, which is what makes the factor "+
			"meaningful: %q", view.IdentityProvider.Note)
	}

	// And no session row can carry a token or a factor value — the type has nowhere to put one.
	encoded, err := json.Marshal(view.Sessions)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	for _, token := range h.tokens {
		if strings.Contains(string(encoded), token) {
			t.Fatal("a session bearer token reached the oversight read model")
		}
	}
}

// TestAnObservabilityIntegrationHasThreeStatesAndNamesItsFailure defends P26 task 6.3.
func TestAnObservabilityIntegrationHasThreeStatesAndNamesItsFailure(t *testing.T) {
	if got := len(adminops.IntegrationStates()); got != 3 {
		t.Fatalf("IntegrationStates() has %d values, want 3 — a boolean would have to call 'not "+
			"configured' a fault or call a fault 'configured'", got)
	}
	h := newHarness(t)
	view := oversightView(t, h, adminops.OversightConfig{Readiness: testReadiness{}})

	if !view.IntegrationsKnown {
		t.Fatal("integrations report as unknown with a readiness source wired")
	}
	byState := map[adminops.IntegrationState]adminops.IntegrationRow{}
	for _, n := range view.Integrations {
		byState[n.State] = n
		if n.Source == "" {
			t.Fatalf("integration %s does not name where its state was read", n.Name)
		}
		// 🔴 Never a third party's dashboard, which is the least available part of the system during an
		// incident — which is exactly when this is read.
		for _, forbidden := range []string{"dashboard", "status page", "console.", ".com"} {
			if strings.Contains(strings.ToLower(n.Source), forbidden) {
				t.Fatalf("integration %s reads its state from %q — reporting health must come from the "+
					"platform's own readiness surface", n.Name, n.Source)
			}
		}
	}
	for _, want := range adminops.IntegrationStates() {
		if _, ok := byState[want]; !ok {
			t.Fatalf("no integration is in state %q — the three states are not all reachable", want)
		}
	}
	if byState[adminops.IntegrationDegraded].FailureClass == "" {
		t.Fatal("a degraded integration names no failure class — 'unreachable' and 'rejecting our " +
			"schema' send an operator to two different places")
	}
	if byState[adminops.IntegrationAbsent].FailureClass != "" {
		t.Fatal("an absent integration names a failure class — nothing has failed; nothing is configured")
	}
}

// TestNoReadinessSourceIsNotReportedAsAbsent defends the distinction the three states exist for.
func TestNoReadinessSourceIsNotReportedAsAbsent(t *testing.T) {
	h := newHarness(t)
	view := oversightView(t, h, adminops.OversightConfig{})

	if view.IntegrationsKnown {
		t.Fatal("integrations report as known with no readiness source")
	}
	if len(view.Integrations) != 0 {
		t.Fatalf("with no readiness source the surface invented %d integration rows. 'Nothing is "+
			"configured' and 'we did not ask' are different answers, and rendering the second as the "+
			"first is a claim nobody made.", len(view.Integrations))
	}
}

// TestAnUnknownDeployedVersionIsStatedAndNamesTheMissingCollection defends P26 task 6.4.
func TestAnUnknownDeployedVersionIsStatedAndNamesTheMissingCollection(t *testing.T) {
	h := newHarness(t)
	view := oversightView(t, h, adminops.OversightConfig{
		Tenants: func() []string { return []string{tenantAcme, tenantBoreal} },
	})

	if len(view.Deployments) != 2 {
		t.Fatalf("the surface shows %d deployments, want 2", len(view.Deployments))
	}
	for _, d := range view.Deployments {
		if !d.Unknown {
			t.Fatalf("tenant %s reports a deployed version with no source — no version may be inferred "+
				"from an API contract version, a feature probe, or any other proxy", d.TenantID)
		}
		if d.Version != "" {
			t.Fatalf("tenant %s reports version %q while also reporting unknown", d.TenantID, d.Version)
		}
		if d.MissingCollection != adminops.MissingDeploymentHeartbeat {
			t.Fatalf("tenant %s's unknown names %q as the missing collection, want %q — an unnamed "+
				"missing input is a wish, not a task", d.TenantID, d.MissingCollection,
				adminops.MissingDeploymentHeartbeat)
		}
	}
}

// TestPaymentsAreStatedAsNotYetReadable defends P26 task 6.5.
//
// 🔴 The failure it prevents is a page that renders an empty payments table and reads as "no failed
// payments" — which is the most dangerous possible reading of a subsystem that does not exist.
func TestPaymentsAreStatedAsNotYetReadable(t *testing.T) {
	h := newHarness(t)
	view := oversightView(t, h, adminops.OversightConfig{})

	var payments *adminops.NotYetReadable
	for i := range view.NotYetReadable {
		if strings.Contains(view.NotYetReadable[i].Subject, "payment") {
			payments = &view.NotYetReadable[i]
		}
	}
	if payments == nil {
		t.Fatal("the payments gap is not stated at all")
	}
	if payments.Requires == "" {
		t.Fatal("the payments gap names no collection — an unnamed missing input is a wish")
	}
	if !strings.Contains(payments.Requires, "webhook") || !strings.Contains(payments.Requires, "dunning") {
		t.Fatalf("the payments gap does not specify the webhook and dunning surface: %q", payments.Requires)
	}
	if !strings.Contains(payments.Statement, "not shipped") {
		t.Fatalf("the statement does not say the subsystem has not shipped: %q", payments.Statement)
	}
	// It must not offer a zero, and it must say so.
	if !strings.Contains(payments.Statement, "nothing below is a zero") {
		t.Fatalf("the statement does not refuse a zero: %q", payments.Statement)
	}
	// And the deployed-version gap is stated in the same shape.
	var sawVersion bool
	for _, n := range view.NotYetReadable {
		if strings.Contains(n.Subject, "deployed version") {
			sawVersion = true
			if n.Requires != adminops.MissingDeploymentHeartbeat {
				t.Fatalf("the deployed-version gap names %q", n.Requires)
			}
		}
	}
	if !sawVersion {
		t.Fatal("the deployed-version gap is not stated")
	}
}

// TestTheOversightSurfaceIsReadOnly defends the phase's boundary on this surface too.
func TestTheOversightSurfaceIsReadOnly(t *testing.T) {
	allowed := map[string]bool{"View": true}
	typ := reflect.TypeOf(&adminops.OversightService{})
	for i := 0; i < typ.NumMethod(); i++ {
		if name := typ.Method(i).Name; !allowed[name] {
			t.Fatalf("OversightService exposes %q — this surface reads and does nothing", name)
		}
	}
	h := newHarness(t)
	if !oversightView(t, h, adminops.OversightConfig{}).ReadOnly {
		t.Fatal("the oversight read model does not declare itself read-only")
	}
}
